### PR TITLE
Interrupt Progress + Progress Action Improvements

### LINKED ISSUE #2204: Progress Actions need to be interrupted in more situations
# Description
Currently they are interrupted if you move. But they aren't interrupted by dropping your in hand item, which can cause glitches in several cases (try constructing a Rack and drop the item before it's complete).

Think of other cases where progress needs to be interrupted and try to add the interrupt call for those. Please add to this list if you think of something. Otherwise just try to implement interrupts for these things:

- [ ] When dropping the item being used to perform the action
- [ ] When moving the item being used to perform the action to a different inventory slot
- [ ] When doing something with bare hands and picking up an item in the active hand
- [ ] When doing something with bare hands and moving an item into the active hand from an inventory slot
- [ ] When downed due to falling over or going crit
- [ ] When pushed or pulled (this might already interrupt)

### PR DESCRIPTION
**Pull request by chairbender**
### Purpose
Fixes #2204 

Refactors progress actions to provide a more usable and flexible API, decoupling the logic of progress from the UI components. It should now be capable of supporting a wider range of use cases without having to rework the internals of progress actions.
Adds an ActionTarget class which should be a useful way to abstract out the differences between actions which target tiles and actions which target specific objects. This is intended to be used for more than just progress actions, for example some interaction logic and other systems which involve targeting something could benefit from not having to distinguish between tile targets and object targets.

### CODE REVIEW

File: UnityProject/Assets/Scripts/UI/ProgressBar/StandardProgressAction.cs [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+			.Where(pb => pb.RegisterPlayer.gameObject == info.Performer)
+			.Where(pb => pb.transform.parent == targetParent)
+			.FirstOrDefault(pb => Vector3.Distance(pb.transform.localPosition, info.Target.TargetLocalPosition) < 0.1);
+		if (existingBar != null)
+		{
+			//progress already started by this player at this position
+			return false;
+		}
+
+		//is this cross matrix? if so, don't start progress if matrix is moving
```
Comment (fomalsd) [inline UnityProject/Assets/Scripts/UI/ProgressBar/StandardProgressAction.cs]: I assume this is temporary?

**Review by fomalsd — APPROVED** [review]
Looking very good code wise! not able to playtest though

File: UnityProject/Assets/Scripts/UI/ProgressBar/StandardProgressAction.cs (hunk 1 above)
Comment (chairbender) [reply UnityProject/Assets/Scripts/UI/ProgressBar/StandardProgressAction.cs]: Not really, there's basically no reason I can think of to allow progress to be started when a crossed matrix is moving.

There is a bug here where it doesn't check if the player's own matrix is moving, though.

**Comment by chairbender** [comment]
I'll fix the cross matrix thing now

**Comment by chairbender** [comment]
good to review again

**Pull request merged**